pkg/services/notary/notary.go

```diff
@@ -328,7 +322,72 @@ func (n *Notary) finalize(tx *transaction.Transaction) error {
 		return fmt.Errorf("failed to update completed transaction's size: %w", err)
 	}
 
-	return n.onTransaction(newTx)
+	n.pushNewTx(newTx, h)
+
+	return nil
+}
+
+type txHashPair struct {
+	tx       *transaction.Transaction
+	mainHash util.Uint256
+}
+
+func (n *Notary) pushNewTx(tx *transaction.Transaction, h util.Uint256) {
+	select {
+	case n.newTxs <- txHashPair{tx, h}:
+	default:
+	}
+}
+
+func (n *Notary) newTxCallbackLoop() {
+	for {
+		select {
+		case tx := <-n.newTxs:
+			isMain := tx.tx.Hash() == tx.mainHash
+
+			n.reqMtx.Lock()
+			r, ok := n.requests[tx.mainHash]
+			if !ok || isMain && (r.isSent || r.minNotValidBefore <= n.Config.Chain.BlockHeight()) {
+				n.reqMtx.Unlock()
+				continue
+			}
+			if !isMain {
+				// Ensure that fallback was not already completed.
+				var isPending bool
+				for _, fb := range r.fallbacks {
+					if fb.Hash() == tx.tx.Hash() {
+						isPending = true
+						break
+					}
+				}
+				if !isPending {
+					n.reqMtx.Unlock()
+					continue
+				}
+			}
+
+			err := n.onTransaction(tx.tx)
+			if err != nil {
+				n.Config.Log.Error("new transaction callback finished with error", zap.Error(err))
+				n.reqMtx.Unlock()
+				continue
+			}
+
+			if isMain {
+				r.isSent = true
+			} else {
+				for i := range r.fallbacks {
+					if r.fallbacks[i].Hash() == tx.tx.Hash() {
+						r.fallbacks = append(r.fallbacks[:i], r.fallbacks[i+1:]...)
+						break
+					}
+				}
```

Comment:
After this `for` loop we should check if all fallback transactions are sent `(len(r.fallbacks)==0)`. If so, then the request should be removed from the pool.